File: util/Fastq.cc
///////////////////////////////////////////////////////////////////////////////
//                   SOFTWARE COPYRIGHT NOTICE AGREEMENT                     //
//       This software and its documentation are copyright (2009) by the     //
//   Broad Institute.  All rights are reserved.  This software is supplied   //
//   without any warranty or guaranteed support whatsoever. The Broad        //
//   Institute is not responsible for its use, misuse, or functionality.     //
///////////////////////////////////////////////////////////////////////////////
/*
 * \file Fastq.cc
 * \author blau
 * \date Mar 5, 2014
 *
 * \brief Converts a Fastq file to a fastb and qualb.
 */
/*#include <cstdio>

#include "MainTools.h"
#include "Basevector.h"
#include "Qualvector.h"
#include "String.h"
#include "system/ProcBuf.h"
#include <fstream>
#include "util/Fastq.h"
*/

#include <istream>
#include "system/System.h"
#include "String.h"
#include "Basevector.h"
#include "Qualvector.h"
#include "PairsManager.h"
#include "util/Fastq.h"

namespace fastq{


// factored out from FastqToFastbQualb.cc at r48873
bool NextFastqEntry( std::istream& is, FastqEntry &entry,
                            vec<size_t> * p_n_amb, const char phred_offset )
{
    bool result = false;
    std::string seqheader;
    std::string seq;
    std::string qualheader;
    std::string qual;
    if ( std::getline(is,seqheader) && seqheader.size() )
    {
        std::getline(is,seq);
        std::getline(is,qualheader);
        std::getline(is,qual);

        // ---- count number of ambiguous bases
        typedef std::string::iterator Itr;
        for ( Itr itr(seq.begin()), end(seq.end()); itr != end; ++itr )
            (*p_n_amb)[GeneralizedBase::fromChar(*itr).getAmbiguityCount()-1]++;

        entry.sequence = bvec(seq.begin(),seq.end(),GenCharToRandomBaseMapper());

        entry.qualities.clear().reserve(qual.size());
        Swizzler swiz(phred_offset);
        for ( Itr itr(qual.begin()), end(qual.end()); itr != end; ++itr )
            entry.qualities.push_back(swiz(*itr));
        entry.name = seqheader;
        entry.name = entry.name.SafeAfter( "@" );
        entry.phred_min = swiz.getMin();
        entry.phred_max = swiz.getMax();
        result = true;
    }
    return result;
}

void Fastq2FastbQualbPair_Interleave( const String& fastq1, const String& fastq2, const char q_offset
                                    , const String& outhead){
    IncrementalWriter<basevector> bases_out((outhead + ".fastb").c_str());
    IncrementalWriter<qualvector> quals_out((outhead + ".qualb").c_str());
    uint64_t nPairs=0;
    FastQReader r1(fastq1,q_offset),r2(fastq2,q_offset);
    for(;r1.State()&&r2.State(); r1.Step(),r2.Step()){
        if(r1.Get().sequence.size() != r1.Get().qualities.size()){ FatalErr("#base != #qual for read 1 of pair "+ToString(nPairs)+"."); }
        if(r2.Get().sequence.size() != r2.Get().qualities.size()){ FatalErr("#base != #qual for read 2 of pair "+ToString(nPairs)+"."); }
        bases_out.add(r1.Get().sequence);
        bases_out.add(r2.Get().sequence);

        quals_out.add(r1.Get().qualities);
        quals_out.add(r2.Get().qualities);
        ++nPairs;
    }
    if(r1.State() != r2.State()){ FatalErr("The two fastq files have different number of entries."); }

    PairsManager pairs(nPairs * 2);
    int lib_id = pairs.addLibrary(0, 0, fastq1+"_"+fastq2);
    for (size_t i = 0; i < nPairs; i++) pairs.addPairToLib( i * 2, i * 2 + 1, lib_id);
    pairs.Write(outhead + ".pairs");
}
void Fastq2FastbQualbPair( const String& fastq, const char q_offset , const String& outhead){
    IncrementalWriter<basevector> bases_out((outhead + ".fastb").c_str());
    IncrementalWriter<qualvector> quals_out((outhead + ".qualb").c_str());
    uint64_t nReads=0;
    for(FastQReader rr(fastq,q_offset) ; rr.State() ; rr.Step()){

        if(rr.Get().sequence.size() != rr.Get().qualities.size()){ FatalErr("#base != #qual for read "+ToString(nReads)+"."); }
        bases_out.add(rr.Get().sequence);
        quals_out.add(rr.Get().qualities);
        ++nReads;
    }
    if(nReads%2!=0) FatalErr("odd number of reads.");

    PairsManager pairs(nReads);
    int lib_id = pairs.addLibrary(0, 0, fastq);
    for (size_t i = 0; i < nReads; i+=2) pairs.addPairToLib( i * 2, i * 2 + 1, lib_id);
    pairs.Write(outhead + ".pairs");
}



size_t ReadFastq( String const& filename, vecbasevector& bases_out,
        vecqualvector& quals_out, char const phred_offset,
        vec<String>* names_outp) {

    bases_out.clear();
    quals_out.clear();

    size_t count = 0;
    for(FastQReader rr(filename,phred_offset) ; rr.State() ; rr.Step()){
        if(rr.Get().sequence.size() != rr.Get().qualities.size()){
            FatalErr("#base != #qual for read "+
                    ToString(count)+" from "+filename); }
        bases_out.push_back(rr.Get().sequence);
        quals_out.push_back(rr.Get().qualities);
        if ( names_outp ) names_outp->push_back(rr.Get().name);
        count++;
    }

    return count;
}


}

File: python/BroadCRD/util/Fastq.py
###############################################################################
#                    SOFTWARE COPYRIGHT NOTICE AGREEMENT                      #
#        This software and its documentation are copyright (2012) by the      #
#    Broad Institute.  All rights are reserved.  This software is supplied    #
#    without any warranty or guaranteed support whatsoever. The Broad         #
#    Institute is not responsible for its use, misuse, or functionality.      #
###############################################################################

# Simple functions for reading and writing FASTQ data.

# Original author: Michael G. Ross <mgross@broadinstitute.org>

class FastqFormatError(Exception):
    def __init__(self, state):
        self.state = state
        return
    def __str__(self):
        return repr(self.state)

def readFastq(filename):
    name_read_quality_triples = []
    with open(filename, 'r') as fastq_data:
        current_name = ''
        current_seq = ''
        current_plus = ''
        current_qual = ''

        for line_num, fastq_line in enumerate(fastq_data):
            if (not current_name and not current_seq and not current_plus and 
                not current_qual and fastq_line[0] == '@'):
                current_name = fastq_line[1:].rstrip()
            elif (current_name and not current_plus and not current_qual and 
                fastq_line[0] != '+'):
                current_seq += fastq_line.rstrip()
            elif (current_name and current_seq and not current_plus and
                not current_qual and fastq_line[0] == '+'):
                current_plus = fastq_line.rstrip()
            elif (current_name and current_seq and current_plus and
                len(current_qual) < len(current_seq)):
                current_qual += fastq_line.rstrip()
            else:
                raise FastqFormatError, '{0}:{1}'.format(filename, line_num)
            if (current_name and current_seq and current_plus and
                len(current_qual) == len(current_seq)):
                name_read_quality_triples.append((current_name, current_seq,
                    current_qual))
                current_name = ''
                current_seq = ''
                current_plus = ''
                current_qual = ''

    return name_read_quality_triples

def writeFastq(name_read_quality_triples, filename):
    with open(filename, 'w') as fastq_out:
        for nrq in name_read_quality_triples:
            print >>fastq_out, '@{0}\n{1}\n+\n{2}\n'.format(nrq[0], nrq[1],
                nrq[2]),
    return
    
def readFastqDict(filename):
    fastq_list = readFastq(filename)
    fastq_dict = {}
    for f in fastq_list:
        fastq_dict[f[0]] = (f[1], f[2])
    return fastq_dict

def writeFastqDict(fastq_dict, filename):
    contig_names = list(fastq_dict.keys())
    contig_names.sort()
    name_read_quality_triples = []
    for c in contig_names:
        (read, qual) = fastq_dict[c]
        name_read_pairs.append((c, read, qual))
    writeFastq(name_read_quality_triples, filename)
    return
